### PR TITLE
Initial framework for auditing the archives

### PR DESCRIPTION
**Pull request by portante**
This is by no means complete, and hopefully will serve as a start for
others to add to significantly.

The state directory contents are not audited at all.

The prefix files are not checked to see that their respective tar balls
exist.

The reception areas are not checked.

### CODE REVIEW

File: server/pbench/bin/pbench-audit-archive [hunk 1]
```diff
@@ -0,0 +1,196 @@
+#! /bin/bash
+
+# Audit the fs-version-001 archive directory structure.
+
+# No changes are made here!
+
+# Algorithm:
+#   For each controller do:
+#     Verify all sub-directories of a given controller are one
+#       of the expect state directories
```
Comment (ndokos) [inline server/pbench/bin/pbench-audit-archive]: expected

File: server/pbench/bin/pbench-audit-archive [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+echo "start-$(timestamp)"
+
+# Initialize index mail content
+> $index_content
+
+# Construct the report file
+> $report
+
+find $ARCHIVE -maxdepth 1 ! -type d -ls | sort -k 11 > $badcontrollers
```
Comment (ndokos) [inline server/pbench/bin/pbench-audit-archive]: Can you add a comment to explain the 11? Maybe even the whole command: "find non-directories in the archive top-level directory and sort them alphabetically by name"  - assuming that's correct.

File: server/pbench/bin/pbench-audit-archive [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+if [ -s $badcontrollers ]; then
+    printf "\n\nBad Controllers:\n"
+    cat $badcontrollers
+fi
+
+find $ARCHIVE -maxdepth 1 -type d ! -name . ! -name $(basename $ARCHIVE) | sort > $controllers
```
Comment (ndokos) [inline server/pbench/bin/pbench-audit-archive]: Ditto - in particulary, what is the "! -name ." protecting us from?

**Comment by portante** [comment]
This PR depends on #943.

**Comment by portante** [comment]
@ndokos, @tenstormavi, this is ready to be considered for review and merging.

File: server/pbench/bin/pbench-audit-archive [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+function verify_prefixes {
+    controller=${1}
+
+    if [ ! -e ${controller}/.prefix ]; then
+        return 0
+    fi
+    if [ ! -d ${controller}/.prefix ]; then
+        printf "\t* Prefix directory, .prefix, is not a directory\n"
```
Comment (ndokos) [inline server/pbench/bin/pbench-audit-archive]: "... ${controller}/.prefix ..." ?

File: server/pbench/bin/pbench-audit-archive [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        printf "\t  ++++++++++\n"
+        cat ${nonprefixes}
+        printf "\t  ----------\n"
+        let cnt=cnt+1
+    fi
+    rm -f ${nonprefixes}
+
+    find ${controller}/.prefix -maxdepth 1 -name 'prefix.*' -printf "\t  %f\n" | sort > ${wrongprefixes} 2>&1
+    if [ -s ${wrongprefixes} ]; then
+        printf "\t* Wrong prefix file names found in /.prefix directory:\n"
```
Comment (ndokos) [inline server/pbench/bin/pbench-audit-archive]: "... ${controller}/.prefix ..."?

File: server/pbench/bin/pbench-audit-archive [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    if [ ! -d ${controller}/.prefix ]; then
+        printf "\t* Prefix directory, .prefix, is not a directory\n"
+        return 1
+    fi
+
+    let cnt=0
+
+    find ${controller}/.prefix -maxdepth 1 ! -name 'prefix.*' ! -name '*.prefix' -printf "\t  %f\n" | sort > ${nonprefixes} 2>&1
+    if [ -s ${nonprefixes} ]; then
+        printf "\t* Unexpected file system objects in .prefix directory:\n"
```
Comment (ndokos) [inline server/pbench/bin/pbench-audit-archive]: "... ${controller}/.prefix ..."?

File: server/pbench/bin/pbench-audit-archive [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+# $ARCHIVE directory itself, while keeping them all in sorted order.
+find $ARCHIVE -maxdepth 1 -type d ! -name . ! -name $(basename $ARCHIVE) | sort > $controllers
+while read controller ;do
+    lclreport="${workdir}/$(basename ${controller})"
+    > $lclreport
+    verify_subdirs $controller >> $lclreport
+    verify_tarball_names $controller >> $lclreport
+    verify_prefixes $controller >> $lclreport
+    if [ -s $lclreport ]; then
+        printf "\n\nController: $(basename ${controller})\n" >> $report
```
Comment (ndokos) [inline server/pbench/bin/pbench-audit-archive]: Maybe this is enough and you can disregard the previous comments.

File: server/pbench/bin/pbench-audit-archive (hunk 4 above)
Comment (portante) [reply server/pbench/bin/pbench-audit-archive]: Line 180 emits the controller name before these reports, so was hoping to not duplicate information if possible.

File: server/pbench/bin/pbench-audit-archive (hunk 7 above)
Comment (portante) [reply server/pbench/bin/pbench-audit-archive]: Was hoping to avoid duplicating information in the report.

**Pull request merged**